### DISCUSSION TITLE
Discrepancy between memtest badram pattern and reported error address?

### DISCUSSION THREAD

**Discussion by nyov** [memtest86plus/memtest86plus#308] [Q&A] [answered]
I'm a bit clueless about computer memory.
Can someone help me understand why these addresses differ? BadRAM reports this address:
![badram](https://github.com/memtest86plus/memtest86plus/assets/438293/c0e1d596-cd44-4a6b-b2d9-42680ffdb4aa)

Switching to the other listing shows a higher Error Address:
![error_address](https://github.com/memtest86plus/memtest86plus/assets/438293/9512a6ad-df2e-4d0a-aa66-14e0c3cc9a4d)
a) I'm not really sure I can trust this report? All the other error listings are empty/flushed when I switch between them during testing, so this one might be corrupt/incomplete as well? Do I need to re-run the whole test? (I ran the test on the badram listing, so I trust that the results in the first photo are complete.)
b) If I can trust this report, I guess I only got a single bad bit there, but why does this address differ from the BadRAM one?

And perhaps, how do I interpret the "Bits in Error Mask" field? Thanks!

**Comment by nyov** [comment]
So I converted the badram pattern into  "`memmap` format" and poked it into the Linux kernel commandline.
I made the region a bit bigger (4KB) and aligned, so it fits kernel the page size (`memmap=0x1000$0xac4d9000` alias `memmap=4K$0xac4d9000`).

`/proc/iomem` shows the region was reserved as requested:
```diff
--- a/iomem
+++ b/iomem
@@ -5,7 +5,9 @@
   000c0000-000cdbff : Video ROM
   000ce000-000cefff : Adapter ROM
 000f0000-000fffff : System ROM
-00100000-bd6f5fff : System RAM
+00100000-ac4d8fff : System RAM
+ac4d9000-ac4d9fff : Reserved
+ac4da000-bd6f5fff : System RAM
 bd6f6000-bd749fff : ACPI Non-volatile Storage
```
I then fired up [`memtester`](https://manpages.debian.org/bullseye/memtester/memtester.8.en.html) to check the reserved region, hoping to find my bit error where I expect it to be:
```
# memtester -p 0xac4d9000
memtester version 4.5.0 (64-bit)
Copyright (C) 2001-2020 Charles Cazabon.
Licensed under the GNU General Public License version 2 (only).

pagesize is 4096
pagesizemask is 0xfffffffffffff000
need memory argument, in MB

Usage: memtester [-p physaddrbase [-d device]] <mem>[B|K|M|G] [loops]
```
However, no such luck:
```
# memtester -p 0xac4d9000 4K 1
memtester version 4.5.0 (64-bit)
Copyright (C) 2001-2020 Charles Cazabon.
Licensed under the GNU General Public License version 2 (only).

pagesize is 4096
pagesizemask is 0xfffffffffffff000
want 0MB (4096 bytes)
Loop 1/1:
  Stuck Address       : ok         
  Random Value        : ok
  Compare XOR         : ok
  Compare SUB         : ok
  Compare MUL         : ok
  Compare DIV         : ok
  Compare OR          : ok
  Compare AND         : ok
  Sequential Increment: ok
  Solid Bits          : ok         
  Block Sequential    : ok         
  Checkerboard        : ok         
  Bit Spread          : ok         
  Bit Flip            : ok         
  Walking Ones        : ok         
  Walking Zeroes      : ok         
  8-bit Writes        : ok
  16-bit Writes       : ok

Done.
```
(I'm only pasting a one-loop run here, however I tested 64 loops, and no errors)

So either `memtester` is unable to see my bad bit here...
or (which I find more likely) my bad bit is hiding in another castle!?

Btw. Memtester86+ found the same badram error in two runs, on two days, with two versions (6.10 and 6.20) so I'm relatively positive that it's not a transient error, and that `memtester` should definitely find something here, if it works.

**Comment by debrouxl** [comment, marked as answer]
Thanks for the report.
Try testing the 0x62c4d96c0 address instead ? I'd venture the guess that the address reported for the badram pattern is (wrongly) a virtual address in the 1 GB test window, instead of a physical address.

**Comment by x86fr** [comment]
Agree...

**Comment by nyov** [comment]
Thanks, I'll go try that other address!

I was still looking into how I got the address wrong, perhaps. But I just stumbled on a post where [someone else mentions that memtest badram patterns may be ...bad?][1]:
> Maybe Memtest86 6.x (EFI computers only) reports badram patterns correctly for x64, older versions don't, neither Memtest86 nor Memtest86+.

But that was posted back in 2015... Surely they would have reported that, or someone else would have stumbled on that in the last ~8 years?
Anyway, I'll go try that other number.

[1]: https://ubuntuforums.org/showthread.php?t=2278744&s=1024c1409acc98fd6bfb97ee5ff8a397&p=13290347#post13290347

**Comment by nyov** [comment]
Amazing! I'll be bleeped. Blasted computers...

The RAM is ded, long live the RAM.

So the kernel commandline this round was `memmap=0x1000$0x000000062c4d9000`

which turns up in `/proc/iomem` as `Reserved` like this:

```diff
--- a/iomem
+++ b/iomem
@@ -99,9 +99,11 @@ c0000000-ffffffff : PCI Bus 0000:00
     fee00000-fee00fff : pnp 00:01
   fef00000-ffffffff : Reserved
     ffc00000-ffffffff : pnp 00:01
-100001000-83effffff : System RAM
-  2af600000-2b0201e47 : Kernel code
-  2b0400000-2b0b63fff : Kernel rodata
-  2b0c00000-2b0e7a0ff : Kernel data
-  2b14b2000-2b25fffff : Kernel bss
+100001000-62c4d8fff : System RAM
+62c4d9000-62c4d9fff : Reserved
+62c4da000-83effffff : System RAM
+  78de00000-78ea01e47 : Kernel code
+  78ec00000-78f363fff : Kernel rodata
+  78f400000-78f67a0ff : Kernel data
+  78fcb2000-790dfffff : Kernel bss
 83f000000-83fffffff : RAM buffer
```

and the test results are:

```
# memtester -p 0x000000062c4d9000 4K 64
memtester version 4.5.0 (64-bit)
Copyright (C) 2001-2020 Charles Cazabon.
Licensed under the GNU General Public License version 2 (only).

pagesize is 4096
pagesizemask is 0xfffffffffffff000
want 0MB (4096 bytes)
Loop 1/64:
  Stuck Address       : testing   0FAILURE: possible bad address line at physical address 0x62c4d96c0.
Skipping to next test...
  Random Value        : ok
  Compare XOR         : ok
  Compare SUB         : ok
  Compare MUL         : ok
  Compare DIV         : ok
  Compare OR          : ok
  Compare AND         : ok
  Sequential Increment: ok
  Solid Bits          : testing  60FAILURE: 0xffffdfffffffffff != 0xffffffffffffffff at physical address 0x62c4d96c0.
  Block Sequential    : ok         
  Checkerboard        : ok         
  Bit Spread          : ok         
  Bit Flip            : testing   2FAILURE: 0xffffdffffffffffe != 0xfffffffffffffffe at physical address 0x62c4d96c0.
  Walking Ones        : ok         
  Walking Zeroes      : ok         
  8-bit Writes        : ok
  16-bit Writes       : ok

[...]

Loop 63/64:
  Stuck Address       : ok         
  Random Value        : ok
  Compare XOR         : ok
  Compare SUB         : ok
  Compare MUL         : ok
  Compare DIV         : ok
  Compare OR          : ok
  Compare AND         : ok
  Sequential Increment: ok
  Solid Bits          : testing  44FAILURE: 0xffffdfffffffffff != 0xffffffffffffffff at physical address 0x62c4d96c0.
  Block Sequential    : ok         
  Checkerboard        : testing  13FAILURE: 0xaaaa8aaaaaaaaaaa != 0xaaaaaaaaaaaaaaaa at physical address 0x62c4d96c0.
  Bit Spread          : ok         
  Bit Flip            : testing 132FAILURE: 0xffffdffffffeffff != 0xfffffffffffeffff at physical address 0x62c4d96c0.
  Walking Ones        : ok         
  Walking Zeroes      : ok         
  8-bit Writes        : ok
  16-bit Writes       : ok

Loop 64/64:
  Stuck Address       : ok         
  Random Value        : ok
  Compare XOR         : ok
  Compare SUB         : ok
  Compare MUL         : ok
  Compare DIV         : ok
  Compare OR          : ok
  Compare AND         : ok
  Sequential Increment: ok
  Solid Bits          : testing  18FAILURE: 0xffffdfffffffffff != 0xffffffffffffffff at physical address 0x62c4d96c0.
  Block Sequential    : ok         
  Checkerboard        : testing  31FAILURE: 0xaaaa8aaaaaaaaaaa != 0xaaaaaaaaaaaaaaaa at physical address 0x62c4d96c0.
  Bit Spread          : testing  84FAILURE: 0x80000000000 != 0x280000000000 at physical address 0x62c4d96c0.
  Bit Flip            : testing   2FAILURE: 0xffffdffffffffffe != 0xfffffffffffffffe at physical address 0x62c4d96c0.
  Walking Ones        : ok         
  Walking Zeroes      : ok         
  8-bit Writes        : ok
  16-bit Writes       : ok

Done.

```

**Comment by x86fr** [comment]
Can you consider donating this know-faulty DRAM stick to Memtest86+? 😅

PS: I'll be happy to send you a know-good DRAM for this stick ;)

**Reply by nyov** [reply]
Well, it seems to be exactly one bad bit, in one 8GB stick, from a matched set of four...  
which is still running my daily driver, holding up for over 10 years now.  
So this is decent RAM... uh, I'll have to think about it. :cold_sweat:

Don't get me wrong, I'm happy to donate something. Just, this particular set is special :D But I suppose any other Stick wouldn't make sense, right?
But finances permitting, I'd be replacing the set with ECC UDIMMs then.

**Reply by x86fr** [reply]
Of course. If you need it, keep it. :) But please think about us if you throw it away in the future :)

**Comment by debrouxl** [comment]
The fact that the lowest error address and highest error address are being reported correctly shows that a defective address is being stored correctly in `app/error.c::error_info`, where it's split into `page` and `offset` by `app/error.c::update_error_info()`. However, the address is not stored correctly in `app/badram.c::patterns`, where it's inserted below `app/badram.c::badram_insert()`, despite the fact that both insertion functions are passed the same `addr` by `app/error.c::common_err()`.

One of the following solutions could be applied:
* `badram_insert()` needs to be passed `page_of((void *)addr) | (addr & 0xFFF)` or equivalently `page_of((void *)addr) + (addr & 0xFFF)`  - single line version of the computation performed by `update_error_info()`;
* this page + offset computation and combination needs to be pushed down into `badram_insert()`;
* this page + offset computation needs to be lifted into `common_err()`, so that `addr` + `page` + `offset` can be passed to `update_error_info()` and used to pass a physical address to `badram_insert()`.

The latter solution would probably scale better if another error reporting method were ever added, so I'd go with this one.

**Comment by x86fr** [comment]
I'm about to rework the error code code a bit for ECC support soon, but this could be solved before.

Solution 1. is a decent fix that requires less testing. 
Solution 3. is probably the better from a technical PoV because both higher level function (`badram_insert()` and `update_error_info()`) should get the correct address directly, but it's also the most intrusive and needs more testing.

PS: Great news, a well-known CPU manufacturer supported us with a low-level JTAG hardware probe. It's able to generate arbitrary memory errors anywhere. This will be extremely helpful for this kind of tests.

**Comment by debrouxl** [comment]
Oh yeah, a JTAG hardware probe will prove invaluable for this kind of tests :)
Therefore, I guess I'll implement the latter solution, it's technically the right one anyway. On my side, it should be testable in the emulated setup I use for working on NUMA support, with a modified test which generates a memory error at some addess.
